File: src/main/java/com/implimentz/fragments/FragmentManager.java
package com.implimentz.fragments;

import android.os.Handler;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Alexander Efremenkov.
 * Date: 04.01.16, 11:31
 * In Intellij IDEA 15.0.1 Ultimate
 * email: implimentz@gmail.com
 * twitter: iamironz
 */
@SuppressWarnings("unchecked")
public final class FragmentManager {

    private final AppCompatActivity activity;
    private final int containerId;
    private final ViewGroup container;
    private final AnnotationManager annotationManager;

    private Toolbar toolbar;
    private StackChangeListener listener;
    private final Handler handler = new Handler();

    public FragmentManager(final AppCompatActivity activity, final int containerId, final Toolbar toolbar) {
        this(activity, containerId);
        this.toolbar = toolbar;
    }

    public FragmentManager(final AppCompatActivity activity, final int containerId) {
        this.activity = activity;
        this.containerId = containerId;
        FragmentHolder.INSTANCE.register(containerId);
        container = (ViewGroup) activity.findViewById(containerId);
        this.annotationManager = new AnnotationManager();
    }

    public final void setStackChangeListener(final StackChangeListener listener) {
        this.listener = listener;
    }

    public final void destroyStack() {
        for (final FragmentData data : getStack()) {
            data.getFragment().onDestroy();
        }

        toolbar = null;
        listener = null;
        FragmentHolder.INSTANCE.unregister(containerId);
    }

    public final void onPause() {
        final List<FragmentData<Object, AppCompatActivity>> stack = getStack();

        if (stack.isEmpty()) {
            return;
        }

        final FragmentData data = stack.get(stack.size() - 1);
        final Fragment fragment = data.getFragment();

        if (!fragment.isShowing()) {
            return;
        }

        fragment.onPause();
    }

    public final void onResume() {
        final List<FragmentData<Object, AppCompatActivity>> stack = getStack();

        if (stack.isEmpty()) {
            return;
        }

        final FragmentData data = stack.get(stack.size() - 1);
        final Fragment fragment = data.getFragment();

        open(fragment);
    }

    public final List<FragmentData<Object, AppCompatActivity>> getStack() {
        return FragmentHolder.INSTANCE.getStackById(containerId);
    }

    public final void openFragment(final String name, final Fragment fragment) {
        openFragment0(name, fragment);
    }

    public final void openFragment(final String name, final Fragment fragment, final long delay) {
        handler.postDelayed(new Runnable() {
            @Override
            public void run() {
                openFragment0(name, fragment);
            }
        }, delay);
    }

    public final void openFragment(final Fragment fragment) {
        openFragment0(fragment.getClass().getName(), fragment);
    }

    public final void openFragment(final Fragment fragment, final long delay) {
        handler.postDelayed(new Runnable() {
            @Override
            public void run() {
                openFragment0(fragment.getClass().getName(), fragment);
            }
        }, delay);
    }

    private void openFragment0(final String name, final Fragment fragment) {
        final List<FragmentData<Object, AppCompatActivity>> stack = getStack();

        if (!stack.isEmpty()) {
            final FragmentData data = stack.get(stack.size() - 1);
            data.getFragment().onPause();
        }

        fragment.setActivity(activity);
        stack.add(new FragmentData(name, fragment));

        open(fragment);
    }

    public final boolean popFragment(final Fragment fragment) {
        return popFragment0(fragment.getClass().getName());
    }

    public final boolean popFragment(final Class<? extends Fragment> fragment) {
        return popFragment0(fragment.getName());
    }

    public final boolean popFragment(final String name) {
        return popFragment0(name);
    }

    private boolean popFragment0(final String name) {
        final List<FragmentData<Object, AppCompatActivity>> stack = getStack();

        if (stack.isEmpty()) {
            return false;
        }

        for (int i = 0; i < stack.size(); i++) {
            final FragmentData data = stack.get(i);
            if (data.getName().equals(name)) {
                closeFragmentRange(i + 1);
                open(data.getFragment());
                return true;
            }
        }

        return false;
    }

    private void closeFragmentRange(final int start) {
        final List<FragmentData<Object, AppCompatActivity>> stack = getStack();

        final List<FragmentData> forDeleting = new ArrayList<>();

        for (int i = start; i < stack.size(); i++) {
            final FragmentData data = stack.get(i);
            close(data.getFragment());
            forDeleting.add(data);
        }

        stack.removeAll(forDeleting);
    }

    public final void closeLastFragment() {
        final List<FragmentData<Object, AppCompatActivity>> stack = getStack();

        if (stack.isEmpty() || stack.size() < 2) {
            return;
        }

        final FragmentData old = stack.remove(stack.size() - 1);
        close(old.getFragment());
        getStack().remove(old);

        final FragmentData actual = stack.get(stack.size() - 1);
        open(actual.getFragment());
    }

    public final boolean hasNotEndedActions() {
        final List<FragmentData<Object, AppCompatActivity>> stack = getStack();

        if (stack.isEmpty()) {
            return false;
        }

        final Fragment fragment = stack.get(stack.size() - 1).getFragment();

        return fragment.hasNotEndedAction();
    }

    public final void onActionEndRequired() {
        final List<FragmentData<Object, AppCompatActivity>> stack = getStack();

        if (stack.isEmpty()) {
            return;
        }

        final Fragment fragment = stack.get(stack.size() - 1).getFragment();

        if (fragment.hasNotEndedAction()) {
            fragment.onActionEndRequired();
        }
    }

    public final void closeFragment(final Fragment name) {
        closeFragment0(name.getClass().getName());
    }

    public final void closeFragment(final Class<? extends Fragment> fragment) {
        closeFragment0(fragment.getName());
    }

    public final void closeFragment(final String name) {
        closeFragment0(name);
    }

    private void closeFragment0(final String name) {
        final List<FragmentData<Object, AppCompatActivity>> stack = getStack();

        if (stack.isEmpty()) {
            return;
        }

        for (FragmentData data : stack) {
            if (data.getName().equals(name)) {
                close(data.getFragment());
                getStack().remove(data);
                tryRestoreLast();
                return;
            }
        }
    }

    private void tryRestoreLast() {

        final List<FragmentData<Object, AppCompatActivity>> stack = getStack();

        if (stack.isEmpty()) {
            return;
        }

        final FragmentData data = stack.get(stack.size() - 1);
        open(data.getFragment());
    }

    private void open(final Fragment fragment) {

        if (fragment.isShowing()) {
            return;
        }

        final View view = fragment.constructView(container);

        setMenu(fragment);

        container.removeAllViews();
        container.addView(view);

        fragment.onResume();

        final FragmentMeta meta = annotationManager.getFragmentMeta(fragment);
        updateToolbar(fragment, meta);
        sendFragmentOpened(fragment, meta);
    }

    private void updateToolbar(final Fragment fragment, final FragmentMeta meta) {
        if(toolbar != null) {
            fragment.setToolbar(toolbar);
            final String title = fragment.getTitle();
            final String subTitle = fragment.getSubTitle();
            toolbar.setSubtitle(subTitle);
            if (title == null) {
                toolbar.setTitle(meta.name());
            } else {
                toolbar.setTitle(title);
            }
        }
    }

    private void sendFragmentOpened(final Fragment fragment, final FragmentMeta meta) {
        if (listener != null) listener.onStackChanged(fragment, meta);
    }

    public final boolean isActionModeEnabled() {
        final List<FragmentData<Object, AppCompatActivity>> stack = getStack();

        if (stack.isEmpty()) {
            return false;
        }

        final Fragment fragment = stack.get(stack.size() - 1).getFragment();

        return fragment.isShowing() && fragment.onActionModeEnabled();

    }

    public final void disableActionMode() {
        final List<FragmentData<Object, AppCompatActivity>> stack = getStack();

        if (stack.isEmpty()) {
            return;
        }

        final Fragment fragment = stack.get(stack.size() - 1).getFragment();

        if(!fragment.isShowing()) {
            return;
        }

        fragment.finishActionMode();
    }

    private void close(final Fragment fragment) {
        if (!fragment.isShowing()) {
            return;
        }

        container.removeAllViews();

        clearMenu();

        fragment.onDestroy();
    }

    private void setMenu(final Fragment fragment) {
        if (toolbar != null) {
            toolbar.getMenu().clear();
            final int resId = fragment.onCreateOptionMenu(toolbar.getMenu());
            if (resId != 0) {
                toolbar.inflateMenu(resId);
                toolbar.setOnMenuItemClickListener(new Toolbar.OnMenuItemClickListener() {
                    @Override
                    public boolean onMenuItemClick(MenuItem item) {
                        return fragment.onOptionsItemSelected(item);
                    }
                });
            }
        }
    }

    private void clearMenu() {
        if (toolbar != null) {
            toolbar.getMenu().clear();
        }
    }

    public final int getStackCount() {
        return getStack().size();
    }
}

File: src/main/java/com/implimentz/fragments/FragmentManager.kt
@file:Suppress("unused")

package com.implimentz.fragments

import android.view.LayoutInflater
import android.view.ViewGroup
import com.implimentz.fragments.annotation.AnnotationManager
import java.io.Serializable
import java.util.*

/**
 * Created by ironz.
 * Date: 20.01.16, 9:32
 * In Intellij IDEA 15.0.3 Ultimate
 * email: implimentz@gmail.com
 * twitter: iamironz
 */

/**
 * Class that represent stack and base stack control with holding in App class
 */
class FragmentManager(private val container: ViewGroup,
                      private val inflater: LayoutInflater,
                      private val listener: StackChangeListener) {

    private val stack = FragmentHolder.getStackById(container.id)

    init {
        FragmentHolder.register(container.id)
    }

    fun destroyStack() {
        stack.forEach { it.onPause() }
        stack.forEach { it.onDestroy() }
        FragmentHolder.unregister(container.id)
    }

    fun onPause() {

        if (stack.isEmpty()) {
            return
        }

        val fragment = stack.last()

        if (fragment.showing.not()) {
            return
        }

        fragment.onPause()
    }

    fun onResume() {

        if (stack.isEmpty()) {
            return
        }

        val fragment = stack.last()

        tryShow(fragment)
    }

    fun getFragments(): MutableList<Fragment<out Serializable>> {
        return stack
    }

    fun openFragment(name: String, fragment: Fragment<out Serializable>) {
        openFragment0(name, fragment)
    }

    fun openFragment(fragment: Fragment<out Serializable>) {
        openFragment0(fragment.javaClass.name, fragment)
    }

    private fun openFragment0(name: String, fragment: Fragment<out Serializable>) {

        if (stack.isEmpty().not()) {
            val last = stack.last()
            last.onPause()
        }

        fragment.name = name

        stack.add(fragment)

        tryShow(fragment)
    }

    fun popFragment(fragment: Fragment<out Serializable>): Boolean {
        return popFragment0(fragment.javaClass.name)
    }

    fun popFragment(fragment: Class<out Fragment<out Serializable>>): Boolean {
        return popFragment0(fragment.name)
    }

    fun popFragment(name: String): Boolean {
        return popFragment0(name)
    }

    private fun popFragment0(name: String): Boolean {

        if (stack.isEmpty()) {
            return false
        }

        for (i in stack.indices) {
            val fragment = stack[i]
            if (fragment.name == name) {
                closeFragmentRange(i.inc())
                tryShow(fragment)
                return true
            }
        }

        return false
    }

    private fun closeFragmentRange(start: Int) {

        val forDeleting = ArrayList<Fragment<out Serializable>>()

        for (i in start..stack.lastIndex) {
            val fragment = stack[i]
            tryClose(fragment)
            forDeleting.add(fragment)
        }

        stack.removeAll(forDeleting)
    }

    fun closeLastFragment() {

        if (stack.isEmpty() || stack.size < 2) {
            return
        }

        val old = stack[stack.lastIndex]
        tryClose(old)
        stack.remove(old)
        tryRestoreLast()
    }

    fun onBackPressed() {

        if (stack.isEmpty() || stack.size < 2) {
            return
        }

        val item = stack.last()
        item.onBackPressed()
    }

    fun hasNotEndedActions(): Boolean {

        if (stack.isEmpty()) {
            return false
        }

        val fragment = stack.last()

        return fragment.hasNotEndedAction()
    }

    fun onActionEndRequired() {

        if (stack.isEmpty()) {
            return
        }

        val fragment = stack.last()

        if (fragment.hasNotEndedAction()) {
            fragment.onActionEndRequired()
        }
    }

    fun closeFragment(name: Fragment<out Serializable>) {
        closeFragment0(name.javaClass.name)
    }

    fun closeFragment(fragment: Class<out Fragment<out Serializable>>) {
        closeFragment0(fragment.name)
    }

    fun closeFragment(name: String) {
        closeFragment0(name)
    }

    private fun closeFragment0(name: String) {

        if (stack.isEmpty()) {
            return
        }

        stack.asReversed()
                .filter {
                    it.name == name
                }
                .forEach {
                    tryClose(it)
                    stack.remove(it)
                    tryRestoreLast()
                }
    }

    private fun tryRestoreLast() {

        if (stack.isEmpty()) {
            return
        }

        val fragment = stack.last()
        tryShow(fragment)
    }

    private fun tryShow(fragment: Fragment<out Serializable>) {

        if (fragment.showing) {
            return
        }

        tryAddViewToFront(fragment)
        callStackListener(fragment)
    }

    private fun tryAddViewToFront(fragment: Fragment<out Serializable>) {
        val view = fragment.constructView(container, inflater)

        if (container.contains(view).not()) {
            container.addView(view)
            container.hidePrevious()
        }

        fragment.onResume()
    }

    private fun callStackListener(fragment: Fragment<out Serializable>) {
        val meta = AnnotationManager.getMetaOrThrow(fragment)
        val layout = AnnotationManager.getLayoutOrThrow(fragment)
        val analytics = AnnotationManager.getAnalyticsOrNull(fragment)
        val menu = AnnotationManager.getMenuOrNull(fragment)
        listener.onStackChanged(fragment, meta, layout, analytics, menu)
    }

    private fun tryClose(fragment: Fragment<out Serializable>) {
        val view = fragment.view
        if (container.contains(view).not()) {
            return
        }

        container.showPrevious()
        container.removeView(view)
        fragment.onPause()

        if(fragment.configurationChanged) {
            return
        }

        fragment.onDestroy()
    }

    fun getStackCount(): Int {
        return stack.size
    }

    fun stackIsEmpty(): Boolean {
        return stack.size < 2
    }
}